File: text/frequency.py
#!/usr/bin/python3

import sys, time, re, collections

def main():
    if len(sys.argv) != 2:
        print('error: must specify file name')
        sys.exit()
    file_name = sys.argv[1]

    # read entire file into memory
    start_time = time.time()
    text = open(file_name).read()
    elapsed = time.time() - start_time
    print('    slurp  %.3f s' % elapsed)

    # render to lower case
    text = text.lower()
    start_time = time.time()
    elapsed = time.time() - start_time
    print('    lowercase  %.3f s' % elapsed)

    # tokenize
    start_time = time.time()
    words = text.split()
    elapsed = time.time() - start_time
    print('    tokenize  %.3f s' % elapsed)

    # delete non-alphabetic characters from the beginning and end of each token
    start_time = time.time()
    junk = re.compile('^\W|\W$')
    words = [ junk.sub('', word) for word in words ]
    elapsed = time.time() - start_time
    print('    clean  %.3f s' % elapsed)

    # count word frequencies
    start_time = time.time()
    frequency = collections.defaultdict(lambda: 0)
    for word in words:
        frequency[word] += 1
    elapsed = time.time() - start_time
    print('    count  %.3f s' % elapsed)

    # make a sorted list of tuples for testing purposes
    if True:
        counts = sorted(list(frequency.items()),
                key=lambda pair: (-pair[1], pair[0]))
        for word, count in counts[:5]:
            print('%d %s' % (count, word))
        print('...')
        for word, count in counts[-5:]:
            print('%d %s' % (count, word))
    
main()

File: text/frequency.rb
#!/usr/bin/ruby2.0


#--- helper functions

def time_block description, &block
	start_time = Time.now.to_f
	block.call
	elapsed = Time.now.to_f - start_time
	puts '    '+description+'  %.3f s' % elapsed
end


def main
	if ARGV.length != 1
		puts 'error: must specify file name'
		exit
	end
	file_name = ARGV[0]

	# read entire file into memory
	text = nil
	time_block 'slurp' do
		text = File.open(file_name).read
	end

	# render to lower case
	time_block 'lowercase' do
		text.downcase!
	end

	# tokenize
	words = nil
	time_block 'tokenize' do
		words = text.split
	end

	# delete non-alphabetic characters from the beginning and end of each token
	junk = Regexp.new(/^\W+|\W+$/)
	time_block 'clean' do
		words.each { |word| word.gsub! junk, '' }
	end

	# count word frequencies
	frequency = Hash.new(0)
	time_block 'count' do
		words.each {|word| frequency[word] += 1 }
	end

	# make a sorted list of tuples for testing purposes
	if true
		counts = frequency.each_pair.sort_by {|word, count| [-count, word] }
		counts[0..4].each {|word, count| puts '%d %s' % [count, word] }
		puts '...'
		counts[-5..-1].each {|word, count| puts '%d %s' % [count, word] }
	end
	
end

main
